drivers/ipcc/ipcc_register.c

```diff
@@ -0,0 +1,207 @@
+ *
+ * Assumptions/Limitations:
+ *   This function should be called only when ipcc_register is run with
+ *   success and all resources in driver instance are properly allocated.
+ *
+ ****************************************************************************/
+
+void ipcc_cleanup(FAR struct ipcc_driver_s *priv)
+{
+#ifdef CONFIG_IPCC_BUFFERED
+  circbuf_uninit(&priv->ipcc->rxbuf);
+  circbuf_uninit(&priv->ipcc->txbuf);
+#endif
+  nxsem_destroy(&priv->ipcc->rxsem);
+  nxsem_destroy(&priv->ipcc->txsem);
+  priv->ipcc->ops.cleanup(priv->ipcc);
+  nxsem_destroy(&priv->exclsem);
+  kmm_free(priv);
+}
+
+/****************************************************************************
+ * Name: ipcc_register
+ *
+ * Description:
+ *   Create and register the IPCC character driver.
+ *
+ *   IPCC is a simple character driver that supports inter processor
+ *   communication.
+ *
+ * Input Parameters:
+ *   ipcc - An instance of the lower half IPCC driver
+ *   chan - IPCC channel. This will be used ad IPCC minor number.
+ *     IPPC will be registered as /dev/ipccN where N is the minor number.
+ *   buflen - Length of tx and rx buffers, 0 for unbuffered communication.
+ *
+ * Returned Value:
+ *   OK if the driver was successfully registered, or negated errno on
+ *   failure.
+ *
+ * Assumptions/Limitations:
+ *   ipcc is already allocated and initialized by architecture code.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_IPCC_BUFFERED
+int ipcc_register(FAR struct ipcc_lower_s *ipcc, size_t rxbuflen,
+                  size_t txbuflen)
+#else
+int ipcc_register(FAR struct ipcc_lower_s *ipcc)
+#endif
+{
+  FAR struct ipcc_driver_s *priv;
+  char devname[DEVNAME_FMTLEN];
+  int ret;
+
+  /* Allocate a IPCC character device structure */
+
+  if ((priv = kmm_zalloc(sizeof(*priv))) == NULL)
+    {
+      return -ENOMEM;
+    }
+
+  /* Link upper and lower driver together */
+
+  priv->ipcc = ipcc;
+
+#ifdef CONFIG_IPCC_BUFFERED
+  /* allocate buffers for reading and writing data to IPCC memory */
+
+  if (rxbuflen)
+    {
+      if ((ret = circbuf_init(&priv->ipcc->rxbuf, NULL, rxbuflen)))
+        {
+          goto error;
+        }
+    }
+
+  if (txbuflen)
+    {
+      if ((ret = circbuf_init(&priv->ipcc->txbuf, NULL, txbuflen)))
+        {
+          goto error;
+        }
+    }
+#endif /* CONFIG_IPCC_BUFFERED */
+
+  /* Create the character device name */
+
+  snprintf(devname, DEVNAME_FMTLEN, DEVNAME_FMT, ipcc->chan);
+  if ((ret = register_driver(devname, &ipcc_fops, 0666, priv)))
+    {
+      goto error;
+    }
+
+  /* nxsem_init can't really fail us if we provide it with valid params */
+
+  nxsem_init(&priv->exclsem, 0, 1);
+  nxsem_init(&priv->ipcc->rxsem, 0, 0);
+  nxsem_init(&priv->ipcc->txsem, 0, 1);
+
+  return OK;
+
+error:
+#ifdef CONFIG_IPCC_BUFFERED
+  circbuf_uninit(&priv->ipcc->rxbuf);
+  circbuf_uninit(&priv->ipcc->txbuf);
+#endif
+  ipcc_cleanup(priv);
+  kmm_free(priv);
```

Comment:
call kmm_free and circbuf_uninit dobule time

Reply:
Oh my... I did miss that one now, didn't I?